src/wp-includes/theme-templates.php

```diff
@@ -0,0 +1,163 @@
+<?php
+
+/**
+ * Generates a unique slug for templates.
+ *
+ * @access private
+ * @since 5.8.0
+ *
+ * @param string $override_slug The filtered value of the slug (starts as `null` from apply_filter).
+ * @param string $slug          The original/un-filtered slug (post_name).
+ * @param int    $post_ID       Post ID.
+ * @param string $post_status   No uniqueness checks are made if the post is still draft or pending.
+ * @param string $post_type     Post type.
+ * @return string The original, desired slug.
+ */
+function wp_filter_wp_template_unique_post_slug( $override_slug, $slug, $post_ID, $post_status, $post_type ) {
+	if ( 'wp_template' !== $post_type ) {
+		return $override_slug;
+	}
+
+	if ( ! $override_slug ) {
+		$override_slug = $slug;
+	}
+
+	// Template slugs must be unique within the same theme.
+	// TODO - Figure out how to update this to work for a multi-theme
+	// environment.  Unfortunately using `get_the_terms` for the 'wp-theme'
+	// term does not work in the case of new entities since is too early in
+	// the process to have been saved to the entity.  So for now we use the
+	// currently activated theme for creation.
```

Comment:
Is it really such a good idea to merge this to core with an unresolved todo note? I have https://github.com/WordPress/gutenberg/pull/31971 open, which basically solves this fully and will eliminate a lot of future headaches. Although it must be thoroughly tested, it's not intended to change any existing behaviour.

Reply:
Today, is the last day for feature freeze so we need a solution in. I think the current solution is decent but your approach in the PR look promising potentially. We can look at it at ease and make the right call for beta1...

Reply:
I don't think the current solution is scalable because taxonomies and slugs offer too many features that are not suitable for this purpose, nor are those systems aware of the constraints that are needed here. For example, the use of `get_the_terms()[0]` is problematic, because if a template somehow has two theme terms, and one is deleted, it can suddenly change theme. That could then trigger a slug conflict that causes the front-end to render differently.

If this doesn't get fixed for 5.8, a potentially complicated and risky database migration will need to be done in a later version, and it would be nice to avoid that.

Reply:
@carlomanf I think Riad's point is that we need to merge this PR today in order to get the template infrastructure into Core, since it's the last possible date (due to feature freeze) for 5.8. A PR like your WordPress/gutenberg#31971 (which looks quite good to me at first glance BTW!) could be considered a bug fix that we can merge even _after_ feature freeze, so it makes it into 5.8.